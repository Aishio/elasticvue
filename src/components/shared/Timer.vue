<template>
  <div class="d-inline-block" title="Auto reload">
    <v-select v-model="timerSetting"
              :items="timerSettings"
              class="mt-0 pt-0 d-inline-block v-select--dense-append"
              hide-details
              item-text="text"
              item-value="value">
      <template slot="selection" slot-scope="{ item, index }">
        <small v-if="item.value">{{ item.text }}</small>
      </template>
    </v-select>
  </div>
</template>

<script>
  import { onBeforeUnmount, ref, watch } from '@vue/composition-api'

  export default {
    name: 'timer',
    props: {
      action: {
        type: Function,
        default: () => {
        }
      },
      timerSettings: {
        type: Array,
        default: () => {
          return [
            { text: 'None', value: null },
            { text: '5s', value: 5 },
            { text: '15s', value: 15 },
            { text: '30s', value: 30 },
            { text: '60s', value: 60 }
          ]
        }
      },
      defaultSetting: {
        type: [Object, Number],
        default: null
      }
    },
    setup (props) {
      const timerSetting = ref(null)
      const intervalID = ref(null)

      const createInterval = () => {
        intervalID.value = setInterval(() => {
          props.action.call()
        }, timerSetting.value * 1000)
      }
      const destroyInterval = () => {
        clearInterval(intervalID.value)
        intervalID.value = null
      }

      watch(timerSetting, newValue => {
        destroyInterval()
        if (newValue) createInterval()
      })

      if (props.defaultSetting && process.env.NODE_ENV !== 'development') {
        timerSetting.value = props.defaultSetting
        props.action.call()
      }

      onBeforeUnmount(destroyInterval)

      return {
        timerSetting
      }
    }
  }
</script>
