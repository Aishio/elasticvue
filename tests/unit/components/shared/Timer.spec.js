import Vue from 'vue'
import Vuetify from 'vuetify'
import Timer from '@/components/shared/Timer'
import { createLocalVue, mount } from '@vue/test-utils'
import '../../mocks/vuetifyMocks'

jest.useFakeTimers()

describe('components/shared/Timer.vue', () => {
  let localVue
  let vuetify

  beforeEach(() => {
    localVue = createLocalVue()
    vuetify = new Vuetify({
      mocks: {
        $vuetify: {
          t: v => v
        }
      }
    })
    Vue.use(Vuetify)
  })

  describe('snapshots', () => {
    it('should render correct default contents', () => {
      const wrapper = mount(Timer, { localVue, vuetify })
      expect(wrapper.element).toMatchSnapshot()
    })

    it('shows the selected setting', () => {
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { defaultSetting: 5 } })
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('action calls', () => {
    it('does not call the action immediately if no defaultSetting is provided', () => {
      const myAction = jest.fn()
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { action: myAction } })

      wrapper.vm.$nextTick(() => {
        expect(myAction).toHaveBeenCalledTimes(0)
      })
    })

    it('calls the action immediately if a defaultSetting is provided', () => {
      const myAction = jest.fn()
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { defaultSetting: 5, action: myAction } })

      wrapper.vm.$nextTick(() => {
        expect(myAction).toHaveBeenCalledTimes(1)
      })
    })

    it('does not call the action periodically if no defaultSetting is provided', () => {
      const myAction = jest.fn()
      const seconds = 5
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { action: myAction } })

      wrapper.vm.$nextTick(() => {
        jest.advanceTimersByTime(seconds * 1000)
        expect(myAction).toHaveBeenCalledTimes(0)
      })
    })

    it('calls the action periodically if a defaultSetting is provided', () => {
      const myAction = jest.fn()
      const seconds = 5
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { defaultSetting: seconds, action: myAction } })

      wrapper.vm.$nextTick(() => {
        jest.advanceTimersByTime(seconds * 1000)
        expect(myAction).toHaveBeenCalledTimes(2)
      })
    })
  })

  describe('interval management', () => {
    it('starts the interval when a defaultSetting is passed', () => {
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { defaultSetting: 5 } })
      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.intervalID).toBeTruthy()
      })
    })

    it('stops the interval when no time is selected', () => {
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { defaultSetting: 5 } })
      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.intervalID).toBeTruthy()
        wrapper.setData({ timerSetting: null })
        expect(wrapper.vm.intervalID).toBeFalsy()
      })
    })

    it('clears the interval when timer is destroyed', () => {
      const wrapper = mount(Timer, { localVue, vuetify, propsData: { defaultSetting: 5 } })
      wrapper.vm.$nextTick(() => {
        wrapper.destroy()
        expect(wrapper.vm.intervalID).toBeFalsy()
      })
    })
  })
})
