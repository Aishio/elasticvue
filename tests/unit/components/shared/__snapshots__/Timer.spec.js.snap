// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/shared/Timer.vue snapshots should render correct default contents 1`] = `
<div
  class="d-inline-block"
  title="Auto reload"
>
  <div
    class="v-input mt-0 pt-0 d-inline-block v-select--dense-append v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select"
  >
    <div
      class="v-input__control"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="list-3"
        class="v-input__slot"
        role="button"
      >
        <div
          class="v-select__slot"
        >
          <div
            class="v-select__selections"
          >
            <input
              aria-readonly="true"
              autocomplete="off"
              id="input-3"
              readonly="readonly"
              type="text"
            />
          </div>
          <div
            class="v-input__append-inner"
          >
            <div
              class="v-input__icon v-input__icon--append"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-menu-down theme--light"
              />
            </div>
          </div>
          <input
            type="hidden"
          />
        </div>
        <div
          class="v-menu"
        >
          <div
            class="v-menu__content theme--light "
            style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/shared/Timer.vue snapshots shows the selected setting 1`] = `
<div
  class="d-inline-block"
  title="Auto reload"
>
  <div
    class="v-input mt-0 pt-0 d-inline-block v-select--dense-append v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select"
  >
    <div
      class="v-input__control"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="list-11"
        class="v-input__slot"
        role="button"
      >
        <div
          class="v-select__slot"
        >
          <div
            class="v-select__selections"
          >
            <small>
              5s
            </small>
            <input
              aria-readonly="true"
              autocomplete="off"
              id="input-11"
              readonly="readonly"
              type="text"
            />
          </div>
          <div
            class="v-input__append-inner"
          >
            <div
              class="v-input__icon v-input__icon--append"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-menu-down theme--light"
              />
            </div>
          </div>
          <input
            type="hidden"
            value="5"
          />
        </div>
        <div
          class="v-menu"
        >
          <div
            class="v-menu__content theme--light "
            style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
